src/org/exist/security/PermissionFactory.java

```diff
@@ -110,40 +111,52 @@ public static Permission getPermission(final SecurityManager sm, final String us
     }
 
     public static void updatePermissions(final DBBroker broker, final XmldbURI pathUri, final ConsumerE<Permission, PermissionDeniedException> permissionModifier) throws PermissionDeniedException {
-        DocumentImpl doc = null;
         final TransactionManager transact = broker.getBrokerPool().getTransactionManager();
         try(final Txn transaction = transact.beginTransaction()) {
-            final Collection collection = broker.openCollection(pathUri, LockMode.WRITE_LOCK);
-            if (collection == null) {
-                doc = broker.getXMLResource(pathUri, LockMode.WRITE_LOCK);
-                if(doc == null) {
-                    transact.abort(transaction);
-                    throw new XPathException("Resource or collection '" + pathUri.toString() + "' does not exist.");
-                }
+            Collection collection = null;
+            try {
+                collection = broker.openCollection(pathUri, LockMode.WRITE_LOCK);
+                if (collection == null) {
+                    DocumentImpl doc = null;
+                    try {
+                        doc = broker.getXMLResource(pathUri, LockMode.WRITE_LOCK);
+                        if (doc == null) {
+                            transact.abort(transaction);
+                            throw new XPathException("Resource or collection '" + pathUri.toString() + "' does not exist.");
+                        }
+
+                        // keep a write lock in the transaction
+                        transaction.acquireLock(doc.getUpdateLock(), LockMode.WRITE_LOCK);
```

Comment:
To avoid moving the ownership of the lock, which is dangerous and the thing this PR tries to solve. I will explain in detail below